File: go/wait-for-it.go
package main

import (
	"fmt"
	"net"
	"os"
	"os/exec"
	"time"

	flags "github.com/jessevdk/go-flags"
)

type Waitfor struct {
	hostport string
	resolve  bool
	quiet    bool
	strict   bool
	timeout  int64
	interval float64
}

func (w Waitfor) check_resolve() bool {
	_, err := net.ResolveTCPAddr("tcp", w.hostport)
	if err == nil {
		return true
	}
	return false
}

func (w Waitfor) check_connect() bool {
	addr, err := net.ResolveTCPAddr("tcp", w.hostport)
	if err != nil {
		return false
	}
	myaddr := new(net.TCPAddr)
	conn, err := net.DialTCP("tcp", myaddr, addr)
	if err != nil {
		return false
	}
	conn.Close()
	return true
}

func (w Waitfor) run_command(cmd []string) bool {
	err := exec.Command(cmd[0], cmd[1:]...).Run()
	if err != nil {
		if !w.quiet {
			println("error", err)
		}
		return false
	}
	return true
}

func (w Waitfor) run() bool {
	if !w.quiet {
		println("wait for", w.hostport, w.timeout, "sec")
	}
	start := time.Now()
	for {
		if w.resolve {
			if w.check_resolve() {
				if !w.quiet {
					println("resolved", w.hostport)
				}
				return true
			}
		} else {
			if w.check_connect() {
				if !w.quiet {
					println("connected", w.hostport)
				}
				return true
			}
		}
		if w.timeout != 0 && time.Now().Sub(start) > time.Duration(int64(time.Second)*w.timeout) {
			if !w.quiet {
				println("timeout", w.hostport)
			}
			return false
		}
		time.Sleep(time.Duration(float64(time.Second) * w.interval))
	}
}

type Options struct {
	Resolve  bool    `long:"resolve" description:"test resolve(no connect)"`
	Quiet    bool    `short:"q" long:"quiet" description:"Don't output any status messages"`
	Strict   bool    `short:"s" long:"strict" description:"Only execute subcommand if the test succeeds"`
	Timeout  int64   `short:"t" long:"timeout" default:"30" description:"Timeout in seconds, zero for no timeout"`
	Interval float64 `short:"i" long:"interval" default:"1.0" description:"interval second"`
	Version  func()  `short:"V" long:"version" description:"Prints version information"`
}

func main() {
	var opts Options
	opts.Version = func() {
		println("wait-for-it 0.1.0")
		os.Exit(0)
	}
	parser := flags.NewParser(&opts, flags.Default)
	parser.Name = "wait-for-it"
	parser.Usage = "[OPTIONS] HOST:PORT"
	args, err := flags.Parse(&opts)
	if err != nil {
		os.Exit(1)
	}
	if len(args) == 0 {
		parser.WriteHelp(os.Stdout)
		return
	}
	hostport := args[0]
	cmd := args[1:]
	if len(cmd) != 0 && cmd[0] == "--" {
		cmd = cmd[1:]
	}
	w := Waitfor{
		hostport: hostport,
		timeout:  opts.Timeout,
		quiet:    opts.Quiet,
		strict:   opts.Strict,
		resolve:  opts.Resolve,
		interval: opts.Interval,
	}
	res := w.run()
	if res || !w.strict {
		if !w.quiet {
			fmt.Printf("run %s\n", cmd)
		}
		if len(cmd) != 0 {
			w.run_command(cmd)
		}
	}
	if res {
		os.Exit(0)
	}
	os.Exit(1)
}

File: cpp/wait-for-it.cpp
#include <boost/array.hpp>
#include <boost/asio.hpp>
#include <boost/program_options.hpp>
#include <chrono>
#include <iostream>
#include <string>
#include <vector>

using namespace std;
using boost::asio::ip::tcp;

class waitfor {
    string hostport;
    bool quiet;

   public:
    waitfor(string hostport, bool quiet) : hostport(hostport), quiet(quiet) {}

    bool check_resolve() {
        boost::asio::io_context io_context;
        tcp::resolver resolver(io_context);
        size_t idx = hostport.find_last_of(":");
        string host = hostport.substr(0, idx);
        string port;
        if (idx != string::npos) {
            port = hostport.substr(idx + 1);
        }
        boost::system::error_code error;
        tcp::resolver::results_type endpoints =
            resolver.resolve(host, port, error);
        if (error) {
            return false;
        }
        return true;
    }

    bool check_connect() {
        boost::asio::io_context io_context;
        tcp::resolver resolver(io_context);
        size_t idx = hostport.find_last_of(":");
        string host = hostport.substr(0, idx);
        string port;
        if (idx != string::npos) {
            port = hostport.substr(idx + 1);
        }
        boost::system::error_code error;
        tcp::resolver::results_type endpoints =
            resolver.resolve(host, port, error);
        if (error) {
            return false;
        }
        tcp::socket socket(io_context);
        boost::asio::connect(socket, endpoints, error);
        socket.close();
        if (error) {
            return false;
        }
        return true;
    }

    void run_command(vector<string> cmd) {
        char *args[cmd.size() + 1];
        args[cmd.size()] = NULL;
        if (!quiet) {
            cout << "run";
            for (vector<string>::size_type i = 0; i < cmd.size(); i++) {
                cout << " " << cmd.at(i);
                args[i] = strdup(cmd.at(i).c_str());
            }
            cout << endl;
        }
        int res = execvp(args[0], args);
        if (res) {
            perror("exec");
        }
    }

    bool run(bool resolve, int timeout, float interval) {
        if (!quiet) {
            cout << "check " << (resolve ? "resolve" : "connect") << " to "
                 << hostport << ", timeout=" << timeout
                 << ", interval=" << interval << endl;
        }
        std::chrono::milliseconds interval_sec(
            static_cast<int>(interval * 1000));
        chrono::time_point<chrono::system_clock> start =
            chrono::system_clock::now();
        while (true) {
            if (resolve) {
                if (check_resolve()) {
                    if (!quiet) {
                        cout << "resolved "
                             << (chrono::system_clock::now() - start).count() /
                                    1000000.0
                             << " sec." << endl;
                    }
                    break;
                }
            } else {
                if (check_connect()) {
                    if (!quiet) {
                        cout << "connected "
                             << (chrono::system_clock::now() - start).count() /
                                    1000000.0
                             << " sec." << endl;
                        ;
                    }
                    break;
                }
            }
            if ((chrono::system_clock::now() - start).count() >
                timeout * 1000000) {
                if (!quiet) {
                    cout << "timeout "
                         << (chrono::system_clock::now() - start).count() /
                                1000000.0
                         << " sec." << endl;
                }
                return false;
            }
            this_thread::sleep_for(interval_sec);
        }
        return true;
    }
};

int main(int argc, char **argv) {
    using namespace boost::program_options;

    options_description optdesc("wait-for-it");
    optdesc.add_options()("help,h", "show this help");
    optdesc.add_options()("strict,s", "strict option");
    optdesc.add_options()("quiet,q", "be quiet option");
    optdesc.add_options()("resolve", "resolve only");
    optdesc.add_options()("version,v", "version option");
    optdesc.add_options()("timeout,t", value<int>()->default_value(30),
                          "timeout value");
    optdesc.add_options()("interval,i", value<float>()->default_value(1.0),
                          "interval option");
    variables_map vm;
    auto const parsed = parse_command_line(argc, argv, optdesc);
    store(parsed, vm);
    notify(vm);
    if (vm.count("help")) {
        cout << optdesc << endl;
        return 0;
    }
    if (vm.count("version")) {
        cout << "wait-for-it 0.1.0" << endl;
        return 0;
    }
    auto args = collect_unrecognized(parsed.options, include_positional);

    if (args.size() != 0) {
        string hostport = args[0];
        vector<string> cmd(args.begin() + 1, args.end());
        waitfor runner(hostport, vm.count("quiet"));
        bool result = runner.run(vm.count("resolve"), vm["timeout"].as<int>(),
                                 vm["interval"].as<float>());
        if (result || !vm.count("strict")) {
            runner.run_command(cmd);
        }
        return result ? 0 : 1;
    }
    return 1;
}
